File: K/submissions/accepted/paul.cpp
#include <algorithm>
#include <iostream>
#include <vector>
using namespace std;

bool solve() {
	int m, n;
	cin >> m >> n;

	vector<vector<int>> adj(n);
	while (m--) {
		int a, b;
		cin >> a >> b;
		a--, b--;
		adj[a].push_back(b);
		adj[b].push_back(a);
	}

	for (auto &v: adj) {
		sort(begin(v), end(v));
		v.erase(unique(begin(v), end(v)), end(v));
	}

	for (int a = 0; a < n; a++) {
		int non_leaf_neighbors = 0;
		for (int b: adj[a]) {
			non_leaf_neighbors += b != a && size(adj[b]) > 1;
		}
		if (non_leaf_neighbors > 2) return false;
	}

	int C = 0;
	vector<int> comp(n,-1);
	for (int a = 0; a < n; a++) if (comp[a] == -1) {
		auto dfs = [&](const auto &self, int b) -> void {
			comp[b] = C;
			for (int c: adj[b]) {
				if (comp[c] == -1) self(self, c);
			}
		};
		dfs(dfs, a);
		C++;
	}

	vector<int> node_count(C), edge_count(C), loop_count(C);
	for (int a = 0; a < n; a++) {
		node_count[comp[a]]++;
		for (int b: adj[a]) {
			edge_count[comp[a]] += a < b;
			loop_count[comp[a]] += a == b;
		}
	}

	int singles = 0, chains = 0, circles = 0;
	for (int k = 0; k < C; k++) {
		if (edge_count[k] + loop_count[k] == 0) singles++;
		else if (edge_count[k] < node_count[k]) chains++;
		else circles++;
	}
	return circles == 0 || (circles == 1 && chains == 0);
}

int main() {
	cout << (solve() ? "possible" : "impossible") << '\n';
}

File: K/submissions/accepted/Paul.java
import java.io.*;
import java.util.*;

public class Paul {
	static class Node {
		int i, comp = -1;
		Set<Node> adj;

		int nonLeafNeighbors() {
			int res = 0;
			for (Node a: adj) {
				if (a.i != i && a.adj.size() > 1) res++;
			}
			return res;
		}

		void dfs(int comp) {
			this.comp = comp;
			for (Node a: adj) {
				if (a.comp == -1) a.dfs(comp);
			}
		}

		public Node(int i) {
			this.i = i;
			adj = new HashSet<Node>();
		}
	}

	static boolean solve() throws IOException {
		BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
		String[] tokens = in.readLine().split(" ");
		int m = Integer.parseInt(tokens[0]), n = Integer.parseInt(tokens[1]);

		Node[] nodes = new Node[n];
		for (int i = 0; i < n; i++) {
			nodes[i] = new Node(i);
		}
		while (m --> 0) {
			tokens = in.readLine().split(" ");
			int a = Integer.parseInt(tokens[0]) - 1;
			int b = Integer.parseInt(tokens[1]) - 1;
			nodes[a].adj.add(nodes[b]);
			nodes[b].adj.add(nodes[a]);
		}

		for (Node a: nodes) {
			if (a.nonLeafNeighbors() > 2) return false;
		}

		int C = 0;
		for (Node a: nodes) if (a.comp == -1) a.dfs(C++);

		int[] nodeCount = new int[C];
		int[] edgeCount = new int[C];
		int[] loopCount = new int[C];
		for (Node a: nodes) {
			nodeCount[a.comp]++;
			for (Node b: a.adj) {
				if (a.i < b.i)  edgeCount[a.comp]++;
				if (a.i == b.i) loopCount[a.comp]++;
			}
		}

		int singles = 0, chains = 0, circles = 0;
		for (int k = 0; k < C; k++) {
			if (edgeCount[k] + loopCount[k] == 0) singles++;
			else if (edgeCount[k] < nodeCount[k]) chains++;
			else circles++;
		}
		return circles == 0 || (circles == 1 && chains == 0);
	}

	public static void main(String[] args) throws IOException {
		System.out.println(solve() ? "possible" : "impossible");
	}
}
